### PR TITLE
refactor: Repository function and types

### PR DESCRIPTION
**Pull request by nguyenvukhang**
### Summary of changes
- retire `namespace` in favour of regular `type` for `InitProps`
- **feat**: `useFindOneByKey`. Takes in a repo and a key and generates a function, for instance `findOneByName`, that then takes one paramater: name, and searches the repo for matches.
- remove all recursive calls in repositories (`UserRepository` should not have `UserRepository(db)` calls within it)
- rename all repository generators with a `get` prefix
- `Repo` object from `environment.ts` in tests will have all the repositories loaded once `setup()` is ran

### Testing
- `yarn ci:test:all`

### CODE REVIEW

File: packages/database/tests/degree/initialize.test.ts [hunk 1]
```diff
@@ -1,13 +1,11 @@
...
-import { setup, teardown } from '../environment'
+import { setup, teardown, repo, t } from '../environment'
 import { Flatten, oneUp } from '../../src/utils'
 
 const dbName = oneUp(__filename)
 const db = getSource(dbName)
-const t: Partial<{ degree: Degree }> = {}
 
 beforeAll(() => setup(db))
 afterAll(() => teardown(db))
```
Comment (nguyenvukhang) [inline packages/database/tests/degree/initialize.test.ts:11]: `import { repo } from '../environment`

the variable `repo` will have loaded repositories ready-to-go after
`beforeAll(() => setup(db))`

Comment (nguyenvukhang) [reply packages/database/tests/degree/initialize.test.ts:11]: can consider splitting `repo` up to each of `ModuleRepository`, `UserRepository`, ... and importing only those needed in the test.

**Comment by nguyenvukhang** [comment]
ready to merge 🚀

**Comment by nguyenvukhang** [comment]
ready to merge 🚀

File: packages/database/tests/user/eligibleModules/no-optional-params.ts [hunk 2]
```diff
@@ -11,14 +11,12 @@ const db = getSource(dbName)
...
 
-const t: Partial<{ user: User }> = {}
-
 it('Saves a user', async () => {
   expect.assertions(1)
   const props = Init.emptyUser
   props.modulesDone.push('CS1101S')
   await container(db, () =>
-    UserRepository(db)
+    getUserRepository(db)
```
Comment (weiseng18) [inline packages/database/tests/user/eligibleModules/no-optional-params.ts:19]: Repo.User?

File: packages/database/tests/user/getPostReqs/add-module-codes.test.ts [hunk 3]
```diff
@@ -24,20 +20,20 @@ it('Gets all post-reqs', async () => {
   const addModuleCodes = ['MA2001']
   // Get post reqs
   const postReqs = await container(db, () =>
-    UserRepository(db).getPostReqs(t.user, addModuleCodes)
+    getUserRepository(db).getPostReqs(t.user, addModuleCodes)
```
Comment (weiseng18) [inline packages/database/tests/user/getPostReqs/add-module-codes.test.ts:23]: Repo.User

File: packages/database/tests/user/initialize.test.ts [hunk 4]
```diff
@@ -18,7 +17,7 @@ describe('User.initialize', () => {
   it('Saves an empty user', async () => {
     expect.assertions(1)
     await container(db, () =>
-      UserRepository(db)
+      getUserRepository(db)
```
Comment (weiseng18) [inline packages/database/tests/user/initialize.test.ts:20]: Repo.User

File: packages/database/tests/user/getPostReqs/no-optional-params.test.ts [hunk 5]
```diff
@@ -35,15 +28,15 @@ it('Gets all post-reqs', async () => {
   // Get post reqs
   expect.assertions(3)
   await container(db, () =>
-    UserRepository(db)
+    getUserRepository(db)
```
Comment (weiseng18) [inline packages/database/tests/user/getPostReqs/no-optional-params.test.ts:31]: Repo.User

File: packages/database/tests/user/getPostReqs/no-optional-params.test.ts [hunk 6]
```diff
@@ -35,15 +28,15 @@ it('Gets all post-reqs', async () => {
...
       .getPostReqs(t.user)
       .then((res) => {
         expect(res).toBeInstanceOf(Array)
         t.postReqsCodes = res.map(Flatten.module)
         return res
       })
       .then(() =>
-        ModuleRepository(db).findOneBy({
+        getModuleRepository(db).findOneBy({
```
Comment (weiseng18) [inline packages/database/tests/user/getPostReqs/no-optional-params.test.ts:39]: Repo.Module

File: packages/database/tests/user/eligibleModules/no-optional-params.ts (hunk 2 above)
Comment (nguyenvukhang) [reply packages/database/tests/user/eligibleModules/no-optional-params.ts:19]: Yep

**Review by weiseng18 — APPROVED** [review]

**Pull request merged**